### PR TITLE
Improvements to path API related to arrays. Fixes #2007 and #2509

### LINKED ISSUE #2007: Array observer delivers incorrect changeRecord.path index after deleting item in array
I register an array observer:

    observers: [
        'sightChanged(sights.*)'
      ],

sightChanged is then called argument changeRecord = {path: 'sights.2.description', ...}. If I remove something from the sights array before index 2, the path 'sights.2.description' remains the same, even though it should now be decremented by one. This results in index out of bound problems:

screen shot 2015-07-01 at 12 27 14 pm

Seems like a bug to me. For the time being, is there a workaround to observing arrays?

### PR DESCRIPTION
**Pull request by kevinpschaaf**
Previously, `get` & `set` expected paths with array _indicies_ (only), whereas `notifyPath` expected paths with collection _keys_ (only).  Furthermore, paths in wildcard `.*` observer info objects always collection _keys_ (rather than indices for arrays), which could not be passed directly to `get` or `set` since they expected indices.

Included in this PR:

* Denote keys with `#` to disambiguate from index
  * `items.3.name` evaluates to `items[3].name`
  * `items.#3.name` evaluates to `Polymer.Collection.get(this.items).getItem('#3').name`
* `get`, `set`, `notifyPath`, and all other public API's that take paths now accept paths with either keys or indices
  * Involves separating public `notifyPath` (which does index->key conversion) from private `_notifyPath` (which expects key-based paths only), and changing all internal use of `notifyPath` to `_notifyPath`
* Added public `notifySplices` API for integration with alternate data observation systems

### CODE REVIEW

**Comment by kevinpschaaf** [comment]
Topics to review:
* Continue storing collection key as `store` index and converting to/from `#`-prefixed string? Or switch to dictionary + running count and store as `#`-prefixed string and eliminate conversion step?  May need to measure.
* `info` arg added to `get` is kinda weird, but saves 2x walk over path parts.  Alternatives?
* `_notifyPathUp` naming alternatives?

File: src/standard/notify-path.html [hunk 1]
```diff
@@ -146,29 +157,56 @@
...
+              // Replace item associated with key in collection
+              coll.setItem(key, value);
+            }
           }
+          // Set value to object at end of path
           prop[last] = value;
+          // Notify observers of path change
           if (!root) {
-            this.notifyPath(parts.join('.'), value);
+            this._notifyPath(parts.join('.'), value);
```
Comment (sorvell) [inline src/standard/notify-path.html]: Why only if not `root`?

File: src/lib/collection.html [hunk 2]
```diff
@@ -46,10 +46,11 @@
       } else {
         this.pmap[item] = key;
       }
-      return key;
+      return '#' + key;
```
Comment (sorvell) [inline src/lib/collection.html]: Storing these including the `#` seems simpler.

File: src/standard/notify-path.html [hunk 3]
```diff
@@ -146,29 +157,56 @@
...
+          // item in collection associated with key for that item
+          if (array) {
             var coll = Polymer.Collection.get(array);
-            var old = prop[last];
-            var key = coll.getKey(old);
-            parts[i] = key;
-            coll.setItem(key, value);
+            if (parseInt(last) == last) {
+              // Dereference index & lookup collection key
+              var old = prop[last];
```
Comment (sorvell) [inline src/standard/notify-path.html]: Couldn't last be a key? If so, this lookup will fail.

File: src/standard/notify-path.html [hunk 4]
```diff
@@ -186,23 +224,42 @@
...
+       *   indices, the index may be used as a dotted part directly
        *   (e.g. `users.12.name` or `['users', 12, 'name']`).
        * @param {Object=} root Root object from which the path is evaluated.
+       * @param {Object=} info Info object; if supplied, a `path` property
+       *   will be added containing path with array indices converted to keys
        * @return {*} Value at the path, or `undefined` if any part of the path
        *   is undefined.
        */
-      get: function(path, root) {
+      get: function(path, root, info) {
```
Comment (sorvell) [inline src/standard/notify-path.html]: Can we factor `info` into another function for internal use?

File: src/standard/notify-path.html [hunk 5]
```diff
@@ -146,29 +157,56 @@
...
-            coll.setItem(key, value);
+            if (parseInt(last) == last) {
+              // Dereference index & lookup collection key
+              var old = prop[last];
+              var key = coll.getKey(old);
+              // Translate array indices to collection keys for path notificaiton
+              parts[i] = key;
+              // Replace item associated with key in collection
+              coll.setItem(key, value);
+            } else if (last[0] == '#') {
```
Comment (sorvell) [inline src/standard/notify-path.html]: Flip the condition since this is the more common case.

Comment (sorvell) [inline src/standard/notify-path.html]: Flip the condition.

File: src/standard/notify-path.html [hunk 6]
```diff
@@ -330,27 +387,67 @@
...
+       *   ordered and must be normalized/merged before notifying).
+      */
+      notifySplices: function(path, splices) {
+        var info = {};
+        var array = this.get(path, this, info);
+        // Notify change to key-based path
+        this._notifySplices(array, info.path, splices);
+      },
+
+      _notifySplices: function(array, path, splices) {
```
Comment (sorvell) [inline src/standard/notify-path.html]: Add note here that path must contain keys not indices.

File: src/standard/notify-path.html [hunk 7]
```diff
@@ -146,29 +157,56 @@
...
-            prop = prop[part];
-            if (array && (parseInt(part) == part)) {
-              parts[i] = Polymer.Collection.get(array).getKey(prop);
+            if (array && part[0] == '#') {
+              // Part was key; lookup item in collection
+              prop = Polymer.Collection.get(array).getItem(part);
+            } else {
+              // Get item from simple property dereference
+              prop = prop[part];
+              if (array && (parseInt(part) == part)) {
```
Comment (samccone) [inline src/standard/notify-path.html]: We might want to specify the radix here, to eliminate an entire class of odd behavior 

screen shot 2015-10-08 at 5 47 35 pm

**Comment by sorvell** [comment]
LGTM!

**Pull request merged**